project_milestone/tests/test_project_milestone.py

```diff
@@ -0,0 +1,43 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo.tests import common
+
+
+class TestProjectMilestone(common.TransactionCase):
+
+    def setUp(self):
+        super().setUp()
+        self.test_project = self.env['project.project'].create({
+            'name': 'TestProject'})
+        self.test_project_milestone_1 = self.env['project.milestone'].create({
+            'name': 'TestMilestone_1',
+            'project_id': self.test_project.id})
+        self.test_project_milestone_2 = self.env['project.milestone'].create({
+            'name': 'TestMilestone_2',
+            'project_id': self.test_project.id})
+        self.test_open_stage = self.env['project.task.type'].create({
+            'name': 'TestOpenStage'})
+        self.test_close_stage = self.env['project.task.type'].create({
+            'name': 'TestCloseStage',
+            'closed': True})
+        self.env['project.task'].create({
+            'name': 'TestTask',
+            'project_id': self.test_project.id,
+            'milestone_id': self.test_project_milestone_1.id,
+            'stage_id': self.test_open_stage.id})
+        self.env['project.task'].create({
+            'name': 'TestTask',
+            'project_id': self.test_project.id,
+            'milestone_id': self.test_project_milestone_1.id,
+            'stage_id': self.test_close_stage.id})
+
+    def test_milestone_sequences(self):
+        milestone1 = self.test_project_milestone_1
+        milestone2 = self.test_project_milestone_2
+
+        self.assertGreater(milestone2.sequence, milestone1.sequence)
+
+    def test_milestone_progress(self):
+        milestone1 = self.test_project_milestone_1
+
+        self.assertEqual(milestone1.progress, 50)

```

Comment:
```suggestion
        self.assertEqual(milestone1.progress, 50)
        
    def test_read_group_milestone_ids(self):
        # Add the default_project_id in the context and prep the arguments
        
        # Call the method to test
        milestone_ids = self._read_group_milestone_ids(milestone_ids, domain, order)
        
        # Test the result
        self.assetEqual(milestone_ids,??)
```

@patrickrwilson Does this help?

Reply:
Another idea:

```suggestion
        self.assertEqual(milestone1.progress, 50)

    def test_read_group_milestone_ids(self):
        """Ensure all project milestones are displayed properly"""
        domain = [("project_id", "=", self.test_project)]
        # Only one milestone displayed when no project is in context
        Task = self.env["project.task"]
        grouped_tasks = Task.search_read(domain, ["name"], ["milestone_id"])
        self.assertEqual(len(grouped_tasks), 1)
        # All project milestones displayed when a project is in context, even if empty
        Task = Task.with_context(default_project_id=self.test_project.id)
        grouped_tasks = Task.search_read(domain, ["name"], ["milestone_id"])
        self.assertEqual(len(grouped_tasks), 2)
```

Reply:
@Yajo I think I understand (mostly) of what you are trying but I couldn't get it to work. But i've got it very close kind of combining yours with @max3903 suggestions. The last thing is i'm not sure why it's not hitting the if statement inside the _read_group_milestone_ids. I do struggle with context so maybe i'm passing the wrong information?

Reply:
@patrickrwilson The if is based on the variable in the context which you don't set in your test so you don't enter that statement. This line from @Yajo sets the variable in the context:
```
Task = Task.with_context(default_project_id=self.test_project.id)
```